python/feathub/metric_stores/metric.py

```diff
@@ -0,0 +1,210 @@
+# Copyright 2022 The FeatHub Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from abc import ABC, abstractmethod
+from datetime import timedelta
+from typing import Optional, Dict, Union
+
+from feathub.common.utils import append_metadata_to_json
+from feathub.feature_views.transforms.sliding_window_transform import (
+    SlidingWindowTransform,
+)
+from feathub.feature_views.transforms.transformation import Transformation
+
+
+class Metric(ABC):
+    """
+    Common interface for all metrics. A metric refers to a statistic of a
+    characteristic of a feature.
+    """
+
+    def __init__(self, metric_type: str, window_size: timedelta = timedelta(seconds=0)):
+        """
+        :param metric_type: The type of the metric.
+        :param window_size: The time range to compute the metric. It should be zero or
+                            a positive time span. If it is zero, the metric will be
+                            computed from all features that have been processed since
+                            the Feathub job is created.
+        """
+        self.metric_type = metric_type
+        self.window_size = window_size
+
+    def get_tags(self) -> Dict:
+        """
+        Gets the tags of this metric for tag-based metric reporters.
+        """
+        return {
+            "window_time": str(self.window_size),
+        }
+
+    @abstractmethod
+    def get_agg_transform(
+        self, feature_name: str, step_size: timedelta
+    ) -> Union[str, Transformation]:
```

Comment:
Would it be more readable to just have this method return `Transformation`?

Note that we use `str` for end-user to specify ExpressionTransform in a simpler way. But it might not be necessary to support the same shortcut for codes that are only used by FeatHub developers.

Reply:
It seems that `get_transform_functions()` still uses `Union[str, Transformation]` in the method signature.

Would it be simpler for this method to return `Sequence[Callable[[str], Transformation]]`?

Reply:
It seems that `get_transform_functions()` still uses `Union[str, Transformation]` in the method signature.

Would it be simpler for this method to return `Sequence[Callable[[str], Transformation]]`?